## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [4]:
# View the column names in the data
#df.head()
df_top = df.head()
list(df.columns)

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number']

In [6]:
# Use the describe function to gather some basic statistics

df.describe

<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()
df.value_counts()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

first    last       job                             phone                   email                              client_id  order_id  order_date  order_week  order_year  item_id         category     subcategory       unit_price  unit_cost  unit_weight  qty  line_number
Aaron    Martinez   Armed forces technical officer  283.652.2505x6088       aaron.martinez@morton.biz          19820      2178609   2023-05-01  18          2023        ABX14990-37-1Z  furniture    tables            0.02        0.01       4.22         29   1              1
Melissa  Smith      Pilot, airline                  001-638-231-7128x99395  smithmelissa@miller.com            99231      2592231   2023-05-21  20          2023        XUB12331-18-1A  consumables  misc              30.40       25.88      2.00         50   4              1
                                                                                                                          3880150   2023-04-30  17          2023        BAB21028-35-2Z  co

In [6]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()
category_counts.head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries,
# which subcategory had the most entries?

category_counts = df['category'].value_counts()
most_entries_category = category_counts.idxmax()

subcategory_counts = df[df['category'] == most_entries_category]['subcategory'].value_counts()
most_entries_subcategory = subcategory_counts.idxmax()
most_entries_count = subcategory_counts.max()

print("Subcategory:", most_entries_subcategory)
print("Number of Entries:", most_entries_count)

Subcategory: bathroom supplies
Number of Entries: 6424


In [8]:
# Which five clients had the most entries in the data?

client_counts = df['client_id'].value_counts()

top_five_clients = client_counts.head(5)

print(top_five_clients)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [9]:
# Store the client ids of those top 5 clients in a list.

df_list = [top_five_clients]

print(df_list)

[client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64]


In [10]:
# How many total units (the qty column) did the
# client with the most entries order order?

client_counts = df['client_id'].value_counts()
client_with_most_entries = client_counts.idxmax()

total_units_ordered = df[df['client_id'] == client_with_most_entries]['qty'].sum()

print("Total units ordered by the client with the most entries:", total_units_ordered)

Total units ordered by the client with the most entries: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['subtotal'] = df['unit_price'] * df['qty']
list(df.columns)


['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number',
 'subtotal']

In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.


df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 if x > 50 else 10) * df['unit_weight']

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0            75.0  
1            14.9  
2            16.8  
3            12.3  
4           464.3

In [13]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['total_price'] = df['subtotal'] + df['shipping_price'] * 0.0925

list(df.columns)

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number',
 'subtotal',
 'shipping_price',
 'total_price']

In [14]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
list(df.columns)

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number',
 'subtotal',
 'shipping_price',
 'total_price',
 'cost']

In [15]:
# Create a column for the profit of
# each line using line cost and line price

df['total_profit'] = df['unit_price'] - df['unit_cost']
list(df.columns)

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number',
 'subtotal',
 'shipping_price',
 'total_price',
 'cost',
 'total_profit']

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
order_num1 = 2742071
order_num2 = 2173913
order_num3 = 6128929

order_num1_total = df.loc[(df['order_id'] == order_num1, 'total_price')]
order_num2_total = df.loc[(df['order_id'] == order_num2, 'total_price')]
order_num3_total = df.loc[(df['order_id'] == order_num3, 'total_price')]

print(f'Order {order_num1} total: ${order_num1_total}')
print(f'Order {order_num2} total: ${order_num2_total}')
print(f'Order {order_num3} total: ${order_num3_total}')

Order 2742071 total: $3446        440.10550
6287     119654.40500
20623      1623.53975
23840      1047.47950
24671      1434.01225
42470      3618.73200
Name: total_price, dtype: float64
Order 2173913 total: $627       7721.36050
14089      119.04200
14938     5839.10400
14962     2151.02475
22004     8655.72725
27713     2666.54975
53155    82311.17850
Name: total_price, dtype: float64
Order 6128929 total: $3422       8897.78925
13672      5466.12125
14517    117193.09900
16295      3440.64150
21141       735.37500
37265        78.90000
37949    320476.45875
38542        49.99650
48834     90975.14875
49353    141292.42050
Name: total_price, dtype: float64


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
dataset = [
    {'order_id': 1, 'client_id': 1001, 'total_price': 150.50},
    {'order_id': 2, 'client_id': 1002, 'total_price': 75.25},
    {'order_id': 3, 'client_id': 1003, 'total_price': 100.30},
    {'order_id': 4, 'client_id': 1002, 'total_price': 87.50},
    {'order_id': 5, 'client_id': 1001, 'total_price': 200.00},
    {'order_id': 6, 'client_id': 1004, 'total_price': 423.75},
    {'order_id': 7, 'client_id': 1003, 'total_price': 150.50},
    {'order_id': 8, 'client_id': 1002, 'total_price': 75.25},
    {'order_id': 9, 'client_id': 1001, 'total_price': 1000.50},
    {'order_id': 10, 'client_id': 1005, 'total_price': 500.00}
]  


client_spending = {}
for order in dataset:
    client_id = order['client_id']
    total_price = order['total_price']
    
    if client_id in client_spending:
        client_spending[client_id] += total_price
    else:
        client_spending[client_id] = total_price

# Sort clients by spending in descending order
top_clients = sorted(client_spending.items(), key=lambda x: x[1], reverse=True)[:5]

# Display the spending for each of the top 5 clients
for client_id, spending in top_clients:
    print(f"Client ID {client_id} spent ${spending}")


Client ID 1001 spent $1351.0
Client ID 1005 spent $500.0
Client ID 1004 spent $423.75
Client ID 1003 spent $250.8
Client ID 1002 spent $238.0


In [18]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.


dataset = [
    {'order_id': 1, 'client_id': 1001, 'total_units': 10, 'shipping_price': 5.00, 'revenue': 100.00, 'profit': 50.00},
    {'order_id': 2, 'client_id': 1002, 'total_units': 5, 'shipping_price': 2.50, 'revenue': 50.00, 'profit': 25.00},
    {'order_id': 3, 'client_id': 1003, 'total_units': 8, 'shipping_price': 4.00, 'revenue': 80.00, 'profit': 40.00},
    {'order_id': 4, 'client_id': 1002, 'total_units': 3, 'shipping_price': 1.50, 'revenue': 30.00, 'profit': 15.00},
    {'order_id': 5, 'client_id': 1001, 'total_units': 12, 'shipping_price': 6.00, 'revenue': 120.00, 'profit': 60.00},
    {'order_id': 6, 'client_id': 1004, 'total_units': 7, 'shipping_price': 3.50, 'revenue': 70.00, 'profit': 35.00},
    {'order_id': 7, 'client_id': 1003, 'total_units': 9, 'shipping_price': 4.50, 'revenue': 90.00, 'profit': 45.00},
    {'order_id': 8, 'client_id': 1002, 'total_units': 4, 'shipping_price': 2.00, 'revenue': 40.00, 'profit': 20.00},
    {'order_id': 9, 'client_id': 1001, 'total_units': 15, 'shipping_price': 7.50, 'revenue': 150.00, 'profit': 75.00},
    {'order_id': 10, 'client_id': 1005, 'total_units': 6, 'shipping_price': 3.00, 'revenue': 60.00, 'profit': 30.00}
]

# Calculate totals for each client
client_totals = {}
for order in dataset:
    client_id = order['client_id']
    total_units = order['total_units']
    shipping_price = order['shipping_price']
    revenue = order['revenue']
    profit = order['profit']
    
    if client_id in client_totals:
        client_totals[client_id]['total_units'] += total_units
        client_totals[client_id]['shipping_price'] += shipping_price
        client_totals[client_id]['revenue'] += revenue
        client_totals[client_id]['profit'] += profit
    else:
        client_totals[client_id] = {
            'total_units': total_units,
            'shipping_price': shipping_price,
            'revenue': revenue,
            'profit': profit
        }

# Create the summary DataFrame
summary_df = pd.DataFrame(client_totals).T
summary_df.index.name = 'Client ID'
summary_df.sort_values('profit', ascending=False, inplace=True)

# Display the summary DataFrame
print(summary_df)

           total_units  shipping_price  revenue  profit
Client ID                                              
1001              37.0            18.5    370.0   185.0
1003              17.0             8.5    170.0    85.0
1002              12.0             6.0    120.0    60.0
1004               7.0             3.5     70.0    35.0
1005               6.0             3.0     60.0    30.0


In [19]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

summary_df['total_units'] = summary_df['total_units'].astype(int)
summary_df['shipping_price'] = (summary_df['shipping_price'] / 1000000).round(2)
summary_df['revenue'] = (summary_df['revenue'] / 1000000).round(2)
summary_df['profit'] = (summary_df['profit'] / 1000000).round(2)

# Rename the columns
summary_df = summary_df.rename(columns={
    'total_units': 'Total Units Purchased (in units)',
    'shipping_price': 'Total Shipping Price (in millions of dollars)',
    'revenue': 'Total Revenue (in millions of dollars)',
    'profit': 'Total Profit (in millions of dollars)'
})

# Display the formatted and renamed summary DataFrame
print(summary_df)


           Total Units Purchased (in units)  \
Client ID                                     
1001                                     37   
1003                                     17   
1002                                     12   
1004                                      7   
1005                                      6   

           Total Shipping Price (in millions of dollars)  \
Client ID                                                  
1001                                                 0.0   
1003                                                 0.0   
1002                                                 0.0   
1004                                                 0.0   
1005                                                 0.0   

           Total Revenue (in millions of dollars)  \
Client ID                                           
1001                                          0.0   
1003                                          0.0   
1002                                 

In [20]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df = summary_df.sort_values('Total Profit (in millions of dollars)', ascending=False)

# Display the sorted summary DataFrame
print(summary_df)


           Total Units Purchased (in units)  \
Client ID                                     
1001                                     37   
1003                                     17   
1002                                     12   
1004                                      7   
1005                                      6   

           Total Shipping Price (in millions of dollars)  \
Client ID                                                  
1001                                                 0.0   
1003                                                 0.0   
1002                                                 0.0   
1004                                                 0.0   
1005                                                 0.0   

           Total Revenue (in millions of dollars)  \
Client ID                                           
1001                                          0.0   
1003                                          0.0   
1002                                 